File: C#/ProteinTranslation/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using System.Linq;

public static class ProteinTranslation
{
    static readonly Dictionary<string, string[]> PROTEINS = new Dictionary<string, string[]>
    {
        { "Methionine", new[] { "AUG" } },
        { "Phenylalanine", new[] { "UUU", "UUC" } },
        { "Leucine", new[] { "UUA", "UUG"} },
        { "Serine", new[] { "UCU", "UCC", "UCA", "UCG"} },
        { "Tyrosine", new[] { "UAU", "UAC"} },
        { "Cysteine", new[] { "UGU", "UGC"} },
        { "Tryptophan", new[] { "UGG"} },
        { "STOP", new[] { "UAA", "UAG", "UGA"} },
    };

    static string FindKeyFromValue(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return "";
        }

        foreach(KeyValuePair<string, string[]> entry in PROTEINS)
        {
            if(entry.Value.Contains(value))
            {
                return entry.Key;
            }
        }

        return "";
    } 

    public static string[] Proteins(string strand)
    {
        if(String.IsNullOrWhiteSpace(strand))
        {
            return [];
        }

        string pattern = @".{1,3}";
        var codons = Regex.Matches(strand.ToUpper(), pattern);

        List<string> proteins = [];

        foreach(Match c in codons)
        {
            var p = FindKeyFromValue(c.Value);

            if (p == "STOP")
            {
                return proteins.ToArray();
            }

            else if (p != "")
            {
                proteins.Add(p);
            }
        }

        return proteins.ToArray();
    }
}

File: TypeScript/protein-translation.ts
interface Proteins {
    [key: string]: string[];
}

const PROTEINS: Proteins = {
    Methionine: ["AUG"],
    Phenylalanine: ["UUU", "UUC"],
    Leucine: ["UUA", "UUG"],
    Serine: ["UCU", "UCC", "UCA", "UCG"],
    Tyrosine: ["UAU", "UAC"],
    Cysteine: ["UGU", "UGC"],
    Tryptophan: ["UGG"],
    STOP: ["UAA", "UAG", "UGA"]
};

export const translate = (rna: string): string[] => {
    if(rna === undefined) {
        return [];
    }

    let codons = rna.toUpperCase().match(/.{1,3}/g);
    let proteins = [];

    for(let c of codons!) {
        let p = Object.keys(PROTEINS).find(key => PROTEINS[key].includes(c));

        if(p === 'STOP') {
            return proteins;
        }

        else if (p !== undefined) {
            proteins.push(p);
        }

        else {
            throw new Error("Invalid codon");
        }
    };

    return proteins;
};
  
let rna = "AUGUUUUCUUAAAUG";
console.log(translate(rna));